src/win/tty.c

```diff
@@ -937,30 +949,109 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
 
 
 int uv_tty_read_stop(uv_tty_t* handle) {
+  INPUT_RECORD record;
+  DWORD written, err;
+
   handle->flags &= ~UV_HANDLE_READING;
   DECREASE_ACTIVE_COUNT(handle->loop, handle);
 
-  /* Cancel raw read */
-  if ((handle->flags & UV_HANDLE_READ_PENDING) &&
-      (handle->flags & UV_HANDLE_TTY_RAW)) {
+  if (!(handle->flags & UV_HANDLE_READ_PENDING)) {
+    return 0;
+  }
+
+  if (handle->flags & UV_HANDLE_TTY_RAW) {
+    /* Cancel raw read */
     /* Write some bullshit event to force the console wait to return. */
-    INPUT_RECORD record;
-    DWORD written;
     memset(&record, 0, sizeof record);
     if (!WriteConsoleInputW(handle->handle, &record, 1, &written)) {
       return GetLastError();
     }
+  } else if (!(handle->flags & UV_HANDLE_CANCELLATION_PENDING)) {
+    /* Cancel line-buffered read if not already pending */
+    err = uv_cancel_read_console(handle);
+    if (err) {
+      return err;
+    }
+    handle->flags |= UV_HANDLE_CANCELLATION_PENDING;
+  }
+
+  return 0;
+}
+
+static int uv_cancel_read_console(uv_tty_t* handle) {
+  HANDLE active_screen_buffer = INVALID_HANDLE_VALUE;
+  CONSOLE_SCREEN_BUFFER_INFO info;
+  BOOL restore_screen_state;
+  INPUT_RECORD record;
+  DWORD written;
+  DWORD err = 0;
+
+  EnterCriticalSection(&uv_read_console_trap_lock);
+  if (!uv_read_console_started) {
+    /* We have managed to trap the other thread before ReadConsole is called */
+    uv_read_console_trap = TRUE;
+    LeaveCriticalSection(&uv_read_console_trap_lock);
+    return 0;
   }
+  LeaveCriticalSection(&uv_read_console_trap_lock);
 
-  /* Cancel line-buffered read */
-  if (handle->tty.rd.read_line_handle != NULL) {
-    /* Closing this handle will cancel the ReadConsole operation */
-    CloseHandle(handle->tty.rd.read_line_handle);
-    handle->tty.rd.read_line_handle = NULL;
+  /* If ReadConsole already returned (the user pressed ENTER),
+     there is nothing to do */
+  if (WaitForSingleObject(uv_read_console_done, 0) == WAIT_OBJECT_0) {
+    return 0;
   }
 
+  /* From this point return should only happen by jumping to the out label
+     to ensure proper cleanup */
 
-  return 0;
+  /* Save screen state before sending the VK_RETURN event */
+  active_screen_buffer = CreateFileA("conout$",
```

Comment:
I don't think so. It's the documented way of getting active screen buffer handle per https://msdn.microsoft.com/en-us/library/windows/desktop/ms682088(v=vs.85).aspx.
In the worst case it will return INVALID_HANDLE_VALUE.